src/service/position-observer-impl.js

```diff
@@ -0,0 +1,413 @@
+            && layoutRectEquals(prePos.positionRect, position.positionRect)
+            && layoutRectEquals(prePos.viewportRect, position.viewportRect)) {
+          // position doesn't change, do nothing.
+          return;
+        }
+        if (layoutRectsOverlap(position.positionRect, position.viewportRect)) {
+          entry.position = position;
+          // Only call handler if entry element overlap with viewport.
+          try {
+            entry.handler(position);
+          } catch (err) {
+            // TODO(@zhouyx, #9208) Throw error. QQQ dev or user?
+          }
+        } else if (entry.position) {
+          // Need to notify that element gets outside viewport
+          // NOTE: This is required for inabox position observer.
+          entry.position = null;
+          position.positionRect = null;
+          entry.handler(position);
+        }
+      },
+    };
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
+      this.startCallback();
+    }
+
+    this.vsync_.measure(() => {
+      this.updateEntryPosition_(entry);
+    });
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element === element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  unobserve(element) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      if (this.entries_[i].element === element) {
+        this.entries_[i].handler = null;
+        this.entries_.splice(i, 1);
+        if (this.entries_.length == 0) {
+          this.stopCallback();
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot unobserve unobserved element');
+  }
+
+  /** @param {!Object} unusedEntry */
+  updateEntryPosition_(unusedEntry) {}
+
+  startCallback() {}
+
+  stopCallback() {}
+
+  increaseFidelityCallback() {}
+
+  decreaseFidelityCallback() {}
+}
+
+export class AmpDocPositionObserver extends AbstractPositionObserver {
+
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+
+    super(ampdoc);
+
+    /** @private {boolean} */
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {number} */
+    this.scrollTimer_ = Date.now();
+
+    /** @private {Array<function()>} */
+    this.unlisteners_ = [];
+  }
+
+  /* @override */
+  startCallback() {
+    // listen to viewport scroll event to help pass determine if need to
+    this.unlisteners_.push(this.viewport_.onScroll(() => {
+      this.inScroll_ = true;
+      this.scrollTimer_ = Date.now();
+      this.schedulePass_();
+      setTimeout(() => {
```

Comment:
This needs to be hoisted. As it is now, every scroll creates a new function instance.